src/util/migrate.cpp

```diff
@@ -959,7 +988,13 @@ void migrate_expr(const exprt &expr, expr2tc &new_expr_ref)
       members.push_back(new_ref);
     }
 
-    constant_array2t *a = new constant_array2t(type, members);
+    expr2t *a;
+    if(
+      (expr.id() == irept::id_constant && expr.type().id() == typet::t_array) ||
+      expr.id() == typet::t_array)
+      a = new constant_array2t(type, members);
+    else
+      a = new constant_vector2t(type, members);
```

Comment:
You can rewrite this as:
expr2tc a;
if(foo)
  a = constant_array2tc(type, members);
else 
  a = constant_vector2tc(type, members);

or you can use a ternary operator.